lib/modules/coverage/contract_source.js

```diff
@@ -0,0 +1,306 @@
+          // same logic below.
+          node.src = node.expression.src;
+          // falls through
+
+        case 'Assignment':
+        case 'ExpressionStatement':
+          coverage.s[node.id] = 0;
+
+          location = this.sourceMapToLocations(node.src);
+          coverage.statementMap[node.id] = location;
+
+          if(!sourceMapToNodeType[node.src]) sourceMapToNodeType[node.src] = [];
+          sourceMapToNodeType[node.src].push({
+            type: 's',
+            id: node.id,
+            body: {loc: coverage.statementMap[node.id]},
+            parent: node._parent
+          });
+
+
+          children = node.expression ? [node.expression] : [];
+          markLocations = [location];
+          break;
+
+        case 'ContractDefinition':
+        case 'SourceUnit':
+          children = node.nodes;
+          break;
+
+        case 'PragmaDirective':
+          // These nodes do nothing, so we move on.
+          break;
+
+        case 'FunctionDefinition':
+          // Istanbul only wants the function definition, not the body, so we're
+          // going to do some fun math here.
+          var functionSourceMap = new SourceMap(node.src);
+          var functionParametersSourceMap = new SourceMap(node.parameters.src);
+
+          var functionDefinitionSourceMap = new SourceMap(
+            functionSourceMap.offset,
+            (functionParametersSourceMap.offset + functionParametersSourceMap.length) - functionSourceMap.offset
+          ).toString();
+
+          var fnName = node.isConstructor ? "(constructor)" : node.name;
+          location = this.sourceMapToLocations(functionDefinitionSourceMap);
+
+          coverage.f[node.id] = 0;
+          coverage.fnMap[node.id] = {
+            name: fnName,
+            line: location.start.line,
+            loc: location
+          };
+
+          // Record function positions.
+          sourceMapToNodeType[node.src] = [{type: 'f', id: node.id, body: coverage.fnMap[node.id]}];
+
+          if(node.body) children = node.body.statements;
+          markLocations = [location];
+          break;
+
+        default:
+          // console.log(`Don't know how to handle node type ${node.nodeType}`);
+          break;
+      }
+
+      nodesRequiringVisiting = nodesRequiringVisiting.concat(children);
+
+      markLocations.forEach((location) => {
+        for(var i = location.start.line; i <= location.end.line; i++) {
+          coverage.l[i] = 0;
+        }
+      });
+
+    } while(nodesRequiringVisiting.length > 0);
+
+    var contractMatches = true;
+    for(var contractName in this.contractBytecode) {
+      var bytecode = this.contractBytecode[contractName];
+
+      // Try to match the contract to the bytecode. If it doesn't,
+      // then we bail.
+      trace.structLogs.forEach((step) => {
+        if(!contractMatches) return;
+
+        if(!bytecode[step.pc]) {
+          // This contract won't match so we bail
+          contractMatches = false;
+          return;
```

Comment:
You can use `.some` or `.every` if you want to escape the loop ahead of time. (`forEach` doesn't end on a `return`)

Reply:
:+1: